flareon/src/static_files.rs

```diff
@@ -0,0 +1,385 @@
+/// impl FlareonApp for ExampleApp {
+///     fn name(&self) -> &str {
+///         "test_app"
+///     }
+///
+///     fn static_files(&self) -> Vec<(String, Bytes)> {
+///         static_files!("admin/admin.css")
+///     }
+/// }
+/// ```
+#[macro_export]
+macro_rules! static_files {
+    ($($path:literal),*) => {
+        vec![$(
+            (
+                $path.to_string(),
+                $crate::__private::Bytes::from_static(
+                    include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/static/", $path))
+                ),
+            )
+        ),*]
+    };
+}
+
+/// Struct representing a collection of static files.
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct StaticFiles {
+    files: HashMap<String, File>,
+}
+
+impl StaticFiles {
+    /// Creates a new `StaticFiles` instance.
+    #[must_use]
+    fn new() -> Self {
+        Self {
+            files: HashMap::new(),
+        }
+    }
+
+    fn add_file(&mut self, path: &str, content: impl Into<Bytes>) {
+        let mime_type = mime_guess::from_path(path).first_or_octet_stream();
+        let file = File::new(content, mime_type);
+        self.files.insert(path.to_string(), file);
+    }
+
+    #[must_use]
+    fn get_file(&self, path: &str) -> Option<&File> {
+        self.files.get(path)
+    }
+}
+
+impl Default for StaticFiles {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct File {
+    content: Bytes,
+    mime_type: mime_guess::Mime,
+}
+
+impl File {
+    #[must_use]
+    fn new(content: impl Into<Bytes>, mime_type: mime_guess::Mime) -> Self {
+        Self {
+            content: content.into(),
+            mime_type,
+        }
+    }
+
+    #[must_use]
+    fn as_response(&self) -> Response {
+        Response::builder()
+            .header(header::CONTENT_TYPE, self.mime_type.to_string())
+            .body(Body::fixed(self.content.clone()))
+            .expect("failed to build static file response")
+    }
+}
+
+/// Middleware for serving static files.
+///
+/// This middleware serves static files defined by the applications by using
+/// the [`FlareonApp::static_files`] trait method. The middleware serves files
+/// from the `/static/` path.
+///
+/// If a request is made to a path starting with `/static/`, the middleware
+/// checks if the file exists in the static files collection. If it does, the
+/// file is served. Otherwise, the request is passed to the inner service.
+#[derive(Debug, Clone)]
+pub struct StaticFilesMiddleware {
+    static_files: Arc<StaticFiles>,
+}
+
+impl StaticFilesMiddleware {
+    /// Creates a new `StaticFilesMiddleware` instance from the application
+    /// context.
+    #[must_use]
+    pub fn from_app_context(app_context: &AppContext) -> Self {
+        let mut static_files = StaticFiles::new();
+
+        for app in app_context.apps() {
+            for (path, content) in app.static_files() {
+                static_files.add_file(&path, content);
```

Comment:
Does this pass the reference to or copies the content of `content` (I'm not sure how this works with embedded data)? If the latter, it's bad because this files can be big. In that case, we might want to pass a pointer or sth similar.

Reply:
`Bytes` is smart and never copies the data (it's either a pointer to `'static` data, or an `Arc`), so we're fine here.